Cython/Compiler/ModuleNode.py

```diff
@@ -107,6 +108,108 @@ def generate_c_code_config(env, options):
     ('__ge__', '__lt__'): (True, '', None),
 }
 
+class SharedUtilityExporter:
+    """
+    Class responsible for generating code that imports and exports shared utility functions.
+
+    Mark the positions where the functions should be called with `call_import_code()`/`call_export_code()`.
+    The function calls and import/export functions are generated when `generate_exporting_functions()`
+    is called. This approach is needed because the list of the shared functions is only known in the later
+    stages of compilation.
+    """
+    def __init__(self, pos, mod_init_subfunction, scope):
+        self.in_shared_utility_module = bool(scope.context.shared_c_file_path)
+        self.using_shared_utility_module = bool(scope.context.shared_utility_qualified_name)
+        self.pos = pos
+        self.scope = scope
+        self.import_code = mod_init_subfunction("Shared function import code")
+        self.export_code = mod_init_subfunction("Shared function export code")
+
+    def has_shared_exports(self, shared_func_definitions: Sequence[Code.SharedFunctionDecl]):
+        return self.in_shared_utility_module and shared_func_definitions
+
+    def has_shared_imports(self, shared_func_definitions: Sequence[Code.SharedFunctionDecl]):
+        return self.using_shared_utility_module and shared_func_definitions
```

Comment:
Alternatively, if you prefer the cleaner `bool` return, we could use this:
```suggestion
    def has_shared_exports(self, shared_func_definitions: Sequence[Code.SharedFunctionDecl]) -> bool:
        return bool(self.in_shared_utility_module and shared_func_definitions)

    def has_shared_imports(self, shared_func_definitions: Sequence[Code.SharedFunctionDecl] -> bool):
        return bool(self.using_shared_utility_module and shared_func_definitions)
```